[Bug] Firefox setFirefoxUserPrefs is broken

It seems as though `setFirefoxUserPrefs` currently does not work as intended. Given the following code:
```java
try (final Playwright playwright = Playwright.create();
             final Browser browser = playwright.firefox().launch(new LaunchOptions()
                     .setHeadless(false)
                     .setFirefoxUserPrefs(Map.of(
                             "privacy.resistFingerprinting", true,
                             "privacy.resistFingerprinting.randomDataOnCanvasExtract", false)))) {
                                 final BrowserContext context = browser.newContext();
                                 final Response response = context.newPage().navigate("https://browserleaks.com/canvas");
                         }
}
```
There is a runtime error:
```
Caused by: com.microsoft.playwright.PlaywrightException: Unexpected map type: java.util.Map<java.lang.String, java.lang.Object>
	at com.microsoft.playwright.impl.Serialization$StringMapSerializer.serialize(Serialization.java:281) ~[classes/:na]
	at com.microsoft.playwright.impl.Serialization$StringMapSerializer.serialize(Serialization.java:1) ~[classes/:na]
	at com.google.gson.internal.bind.TreeTypeAdapter.write(TreeTypeAdapter.java:81) ~[gson-2.8.6.jar:na]
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.write(TypeAdapterRuntimeTypeWrapper.java:69) ~[gson-2.8.6.jar:na]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.write(ReflectiveTypeAdapterFactory.java:127) ~[gson-2.8.6.jar:na]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.write(ReflectiveTypeAdapterFactory.java:245) ~[gson-2.8.6.jar:na]
	at com.google.gson.Gson.toJson(Gson.java:704) ~[gson-2.8.6.jar:na]
	at com.google.gson.Gson.toJsonTree(Gson.java:597) ~[gson-2.8.6.jar:na]
	at com.google.gson.Gson.toJsonTree(Gson.java:576) ~[gson-2.8.6.jar:na]
	at com.microsoft.playwright.impl.BrowserTypeImpl.launchImpl(BrowserTypeImpl.java:48) ~[classes/:na]
	at com.microsoft.playwright.impl.BrowserTypeImpl.lambda$0(BrowserTypeImpl.java:41) ~[classes/:na]
	at com.microsoft.playwright.impl.LoggingSupport.withLogging(LoggingSupport.java:47) ~[classes/:na]
	at com.microsoft.playwright.impl.BrowserTypeImpl.launch(BrowserTypeImpl.java:41) ~[classes/:na]
	at com.microsoft.playwright.impl.BrowserTypeImpl.launch(BrowserTypeImpl.java:1) ~[classes/:na]
```

When manually patching the latest SNAPSHOT version to fix above runtime exception, the provided user preferences do not seem to be applied to Firefox (when checking `about:config` in the headful window).

--- a/playwright/src/main/java/com/microsoft/playwright/impl/Serialization.java
+++ b/playwright/src/main/java/com/microsoft/playwright/impl/Serialization.java
@@ -48,7 +48,8 @@ class Serialization {
         .registerTypeAdapter((new TypeToken<List<KeyboardModifier>>(){}).getType(), new KeyboardModifiersSerializer())
         .registerTypeAdapter(Optional.class, new OptionalSerializer())
         .registerTypeHierarchyAdapter(JSHandleImpl.class, new HandleSerializer())
-        .registerTypeHierarchyAdapter(Map.class, new StringMapSerializer())
+        .registerTypeAdapter((new TypeToken<Map<String, String>>(){}).getType(), new StringMapSerializer())
+        .registerTypeAdapter((new TypeToken<Map<String, Object>>(){}).getType(), new FirefoxUserPrefsSerializer())
         .registerTypeAdapter(Path.class, new PathSerializer()).create();
     }
     return gson;
@@ -271,6 +272,16 @@ class Serialization {
       JsonObject json = new JsonObject();
       json.addProperty("guid", src.guid);
       return json;
+    }
+  }
+
+  private static class FirefoxUserPrefsSerializer implements JsonSerializer<Map<String, Object>> {
+    @Override
+    public JsonElement serialize(Map<String, Object> src, Type typeOfSrc, JsonSerializationContext context) {
+      if (!"java.util.Map<java.lang.String, java.lang.Object>".equals(typeOfSrc.getTypeName())) {
+        throw new PlaywrightException("Unexpected map type: " + typeOfSrc);
+      }
+      return context.serialize(src, Map.class);
     }
   }
 

--- a/playwright/src/test/java/com/microsoft/playwright/TestBase.java
+++ b/playwright/src/test/java/com/microsoft/playwright/TestBase.java
@@ -23,6 +23,8 @@ import org.junit.jupiter.api.BeforeEach;
 
 import java.io.IOException;
 
+import static com.microsoft.playwright.Utils.getBrowserNameFromEnv;
+
 public class TestBase {
   static Server server;
   static Server httpsServer;
@@ -40,15 +42,15 @@ public class TestBase {
   }
 
   static boolean isChromium() {
-    return "chromium".equals(browserType.name());
+    return "chromium".equals(getBrowserNameFromEnv());
   }
 
   static boolean isWebKit() {
-    return "webkit".equals(browserType.name());
+    return "webkit".equals(getBrowserNameFromEnv());
   }
 
   static boolean isFirefox() {
-    return "firefox".equals(browserType.name());
+    return "firefox".equals(getBrowserNameFromEnv());
   }
 
   private static String getBrowserChannelFromEnv() {

--- a/playwright/src/test/java/com/microsoft/playwright/TestFirefoxLauncher.java
+++ b/playwright/src/test/java/com/microsoft/playwright/TestFirefoxLauncher.java
@@ -1,0 +1,56 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.playwright;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.condition.EnabledIf;
+
+import static com.microsoft.playwright.Utils.mapOf;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class TestFirefoxLauncher extends TestBase {
+
+  @BeforeAll
+  // Hide base class method to not launch browser.
+  static void launchBrowser() {
+  }
+
+  @Override
+  void createContextAndPage() {
+    // Do nothing
+  }
+
+  @Test
+  @EnabledIf(value="com.microsoft.playwright.TestBase#isFirefox", disabledReason="skip")
+  void shouldPassFirefoxUserPreferences() {
+    BrowserType.LaunchOptions options = new BrowserType.LaunchOptions().setFirefoxUserPrefs(
+      mapOf(
+        "network.proxy.type", 1,
+        "network.proxy.http", "127.0.0.1",
+        "network.proxy.http_port", 3333));
+    launchBrowser(options);
+    Page page = browser.newPage();
+    try {
+      page.navigate("http://example.com");
+      fail("did not throw");
+    } catch (PlaywrightException e) {
+      assertTrue(e.getMessage().contains("NS_ERROR_PROXY_CONNECTION_REFUSED"));
+    }
+  }
+}

--- a/playwright/src/test/java/com/microsoft/playwright/Utils.java
+++ b/playwright/src/test/java/com/microsoft/playwright/Utils.java
@@ -104,13 +104,16 @@ class Utils {
     }
   }
 
-  static BrowserType getBrowserTypeFromEnv(Playwright playwright) {
+  static String getBrowserNameFromEnv() {
     String browserName = System.getenv("BROWSER");
-
     if (browserName == null) {
       browserName = "chromium";
     }
+    return browserName;
+  }
 
+  static BrowserType getBrowserTypeFromEnv(Playwright playwright) {
+    String browserName = getBrowserNameFromEnv();
     switch (browserName) {
       case "webkit":
         return playwright.webkit();

